r/src/tpch-gen.cpp

```diff
@@ -0,0 +1,167 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "./arrow_types.h"
+
+#if defined(ARROW_R_WITH_ARROW)
+
+#include <arrow/compute/api.h>
+#include <arrow/compute/exec/exec_plan.h>
+#include <arrow/compute/exec/tpch_node.h>
+#if defined(ARROW_R_WITH_DATASET)
+#include <arrow/dataset/api.h>
+#include <arrow/dataset/plan.h>
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/localfs.h>
+
+namespace compute = ::arrow::compute;
+namespace ds = ::arrow::dataset;
+namespace fs = ::arrow::fs;
+
+static std::shared_ptr<compute::ExecNode> MakeExecNodeOrStop(
+    const std::string& factory_name, compute::ExecPlan* plan,
+    std::vector<compute::ExecNode*> inputs, const compute::ExecNodeOptions& options) {
+  return std::shared_ptr<compute::ExecNode>(
+      ValueOrStop(compute::MakeExecNode(factory_name, plan, std::move(inputs), options)),
+      [](...) {
+        // empty destructor: ExecNode lifetime is managed by an ExecPlan
+      });
+}
```

Comment:
I basically copied this over from `r/src/compute-exec.cpp` which isn't the right thing to do here. I tried adding it to a header but then it complained about a duplicate symbol in this file and copmute-exec What's the appropriate way to share a function like this across two cpp files?

Or I could put all of this into `r/src/compute-exec.cpp` and just gate the includes (since that file doesn't need to require all of the datasets and filesystem stuff to do most of its job?

Reply:
Yeah usually you'd put it in a header file. I don't know why it would complain about a duplicate symbol unless you maybe forgot to delete the old implementation? I guess I'd have to see the specifics more.

Reply:
My guess is you want `#pragma once` at the top of the header file (it's harmless, and maybe a good practice, to add this to the top of every header file actually)